core/lib/pubs.db.test.ts

```diff
@@ -0,0 +1,266 @@
+
+	it("should filter out pub values if needed", async () => {
+		const { seedCommunity } = await import("~/prisma/seed/seedCommunity");
+		const { pubs, community, pubTypes } = await seedCommunity(seed);
+		const { getPubByForm } = await import("./pubs");
+		const { getPubsWithRelatedValues } = await import("./server/pub");
+		const { getForm } = await import("./server/form");
+
+		const [pub, form] = await Promise.all([
+			getPubsWithRelatedValues(
+				{ pubId: pubs[0].id, communityId: community.id },
+				{ withRelatedPubs: true, withMembers: true, withPubType: true, withStage: true }
+			),
+			// Get the default form for this pub
+			getForm({
+				pubTypeId: pubs[0].pubTypeId,
+				communityId: community.id,
+			}).executeTakeFirstOrThrow(),
+		]);
+
+		const pubWithForm = getPubByForm({ pub, form, withExtraPubValues: false });
+		expect(pubWithForm).toMatchObject({
+			values: [
+				// Pub value and form element
+				{
+					value: "Some title",
+					formElementConfig: { label: "Title" },
+				},
+				// Only form element
+				{
+					id: null,
+					value: null,
+					formElementConfig: {
+						relationshipConfig: { label: "Some relation", component: "relationBlock" },
+					},
+				},
+			],
+		});
+	});
+
+	it("should filter out pub values by user permissions", async () => {
+		const trx = getTrx();
+		const { seedCommunity } = await import("~/prisma/seed/seedCommunity");
+		const { pubs, community, users } = await seedCommunity(seed);
+		const { getPubByForm } = await import("./pubs");
+		const { getPubsWithRelatedValues } = await import("./server/pub");
+		const { getForm } = await import("./server/form");
+		const { userCan } = await import("./authorization/capabilities");
+
+		const [pub, form, adminCan, editorCan, contributorCan] = await Promise.all([
+			getPubsWithRelatedValues(
+				{ pubId: pubs[0].id, communityId: community.id },
+				{ withRelatedPubs: true, withMembers: true, withPubType: true, withStage: true }
+			),
+			// Get the default form for this pub
+			getForm({
+				pubTypeId: pubs[0].pubTypeId,
+				communityId: community.id,
+			}).executeTakeFirstOrThrow(),
+			userCan(
+				Capabilities.seeExtraPubValues,
+				{ type: MembershipType.community, communityId: community.id },
+				users.admin.id
+			),
+			userCan(
+				Capabilities.seeExtraPubValues,
+				{ type: MembershipType.community, communityId: community.id },
+				users.editor.id
+			),
+			userCan(
+				Capabilities.seeExtraPubValues,
+				{ type: MembershipType.community, communityId: community.id },
+				users.contributor.id
+			),
+		]);
+
+		[
+			{ withExtraPubValues: adminCan, expected: 3 },
+			{ withExtraPubValues: editorCan, expected: 3 },
+			{ withExtraPubValues: contributorCan, expected: 2 },
+		].forEach(({ withExtraPubValues, expected }) => {
+			const pubWithForm = getPubByForm({ pub, form, withExtraPubValues });
+			expect(pubWithForm.values).toHaveLength(expected);
+		});
+	});
+
+	it("should respect form order", async () => {
```

Comment:
I wanted to add another test that makes sure the pub values not in the form still stay in order, but I'm not sure how to do this in a transaction. I think @kalilsn may have run into this problem too, was there a good workaround?

Reply:
oh that's a good point! I had been trying to control the insertion time so that I could say the order was a static value, but you're right that I should just be able to compare with whatever `getPubsWithRelatedValues` returns to get the same result!